Completely removes `proc_holders` from existence. Refactors all wizard, xeno, spider, and genetics powers to be actions. Also refactors and sorts ton of accompanying code.

MrMelbert opened the pull request:
## About The Pull Request

![image](https://user-images.githubusercontent.com/51863163/168932971-3d0e297c-b43a-4a74-9ecf-baba5292bff9.png)

- Completely removes `/obj/effect/proc_holder`s from existence.
  - What is a proc_holder? 
     - A proc_holder is an ancient method of adding an ability to a mob. They were created before we had action datums.
  - Why are you removing them?
     - Proc holders are atrocious to work with. They have tons and tons of weird random checks and are completely nonsensical when working with them.
     - Proc holders also have no reason to be effects. They are, in every sense of their existence, datums. They are abstract behavior holders that are attached to mobs.

- Refactors all wizard spells to use actions, and sorted their files. 
   - Implements all spells as subtypes of `/datum/action/cooldown`. 
   - This has great added benefit in being able to organize them better. Instead of all spells vaguely being lumped into random subtypes ("targeted"? "aoe_turf"? "pointed"? What does that mean?), they're now grouped in more logical subtypes. 
     - Spells which you click on someone are pointed spells.
     - Spells which shoot a projectile at your mouse cursor are pointed projectile spells.
     - Spells which affect an aoe around you are aoe spells. 
     - Spells which shoot out in a cone are cone spells. 
     - Spells which create one or multiple items in an area around you (or beneath your feet) are conjure spells.
     - Spells which create one item in your hand are conjure item spells.
     - Spells which allow you to pick from a list of targets are list_target spells. 
     - All jaunts are now jaunt spells. 
        - Blood crawl was completely refactored to use actions instead of a bunch of mob level procs and vars. 
        - Jaunter dummies no longer have side effects on destroy(), and is all handled by jaunt spells.
     - All shapeshift spells now share a common parent. 
     - All teleport-type spells now share a common parent, teleport, which divides into two other types - radius turf, which teleports you somewhere in a radius, and area teleport, which teleports you to an entirely new area. 
        - The wizard's teleport scroll now uses the actual teleport spell. 
     - All touch spells are touch spells, as before, but the actual effects of the spell are on the spell itself instead of on the hand. 
        - You can now drop the hand to immediately cancel touch spells. 

- Refactors all xeno abilities to use actions. 
   - Corrosive acid is a click-target action type now, instead of using a weird as heck list. 

- Refactors all spider abilities to use actions.
   - Moved a ton of vars on spider onto the actions themselves. Not entirely done, but in the future actions can be further encapsulated. 

- Refactors all revenant spells to use actions. 

- A few malf AI abilities have been swapped over to actions as well.

- Refactors granters slightly, and sorted their files. 
   - Instead of "onuse" and "used", it now just tracks "uses". 
   - Spell granters are subtypes of action granters. 
   - A few recoil effects which were broken should work now. 
   - Added a "true random" spellbook, just cause. 

- Refactored genetics mutation abilites / spells as well.
   - Genetic "powers" have had a unique spell background set for AGES which never worked. It works now. 

- Did some minor code improvements for action datums, and sorted their files. 

- Refactored the spellbook largely, and sorted their files. 

- Refactored wizards themselves slightly. The roundend report now better lists the spells which the wizard purchased. 

- Refactored how cult's blood target is shown, added, and removed. 

- Unit tests some spells. 

TODO:

- [x] Remove SDQL spells. The creator gave me the green light to do this, as their purposes are being eclipsed by admin circuits and LUA scripting.
- [ ] LOTS and lots of testing. 
- [x] I accidentally removed the hallucination spell from cult and need to re-add it
- [ ] I may have broken construct actions again
- [x] ~~Stat panel entries don't work~~ Turns out the stat panel refactor broke it 

Fixes #67055 ultimately
Fixes #66624
Fixes #45518
Fixes #55789 
Probably fixes #48168 
Probably fixes #55631 

## Why It's Good For The Game

Proc holders are incredibly limited and difficult to work with. 
They're also somewhat buggy and definitely outdated.

By removing them and rolling them into the action datum system, there is a much much greater potential for future spells and abilities. 

## Changelog

:cl: Melbert
add: The Mjolnir wizard loadout no longer scams you. It now also grants 1 level of Tesla Blast!
fix: Spells show up in the stat panel again
fix: Fixes runtime errors that can occur if a wizard spellbook randomizes into unpurchasable spells. 
fix: Genetics powers now have their own background, which apparently they have always meant to have. 
fix: Fixes some cases where chaplains are unable to cast their own spells. 
fix: Fixes some granters being infinite when they should've been limited (Maint loot books). 
fix: Wizard's roundend report is now more accurate in what spells they purchased.
qol: Some more spells and actions now use balloon alerts. 
qol: Many actions will now more accurately have their icon redded out if you can't use them at the moment. 
qol: Genetics mindread is now a pointed ability. 
qol: The mimery book now grants you a vow of silence action, like the guide to advanced mimery. 
qol: Xenomorph's corrosive acid ability is now a click-to-work action, so you click on what you want to acid. 
qol: Also, xenomorph larva now use a radial when choosing an evolution type. 
qol: Summon Item now takes a few seconds to remove the mark from an item, so you no longer accidentally remove it. 
balance: The mindswap book now gives a slight message warning it's changed on read. 
balance: Sacred Flame now lets people around them know they're actually being made super flammable. 
balance: Olfaction can't be cast if you have no nose (you're headless).
balance: Being Emotemute now prevents you from invoking spells which require emotes (mime spells). 
balance: All wizard spells except Mindswap can't be cast if the wizard is just a brain. Most spells didn't work, anyways.
balance: PAIs and carded AIs can no longer cast spells. Most spells didn't work, anyways. 
balance: You can't mindswap into PAIs and brains(?).
admin: Admins can now add spells by name or typepath.
admin: Spell requirements are bitflags now, so you can modify them with the bit-field VV ui. Pretty neat. 
code: proc_holders have been removed from existence. 
refactor: Wizard, Heretic, Revenant, and Genetics spells are now all action datums.
refactor: Xenomorph, Spider, and Cytology monster abilities are now all action datums. 
refactor: Some malf AI and cult abilities are now fully action datums instead of an unholy mix of proc holder and action. 
refactor: Refactored how cult's blood target is set. 
refactor: Bloodcrawling and jaunting have been completely refactored. Blood crawling has been exorcised from the living level of mobs, and the consumption part of (s)laughter demons is now handled by the action itself. 
refactor: The wizard's teleport scroll now actually holds and casts a version of the "teleport" spell, instead of fake casting it. 
refactor: Refactored the wizard spellbook a bit. 
refactor: Refactored how all item actions are applied and handled. 
del: SDQL spells have been removed. 
/:cl:

MrMelbert committed:
destroy proc holder pt1
- change proc_holder/spell to action/cooldown/spell
- docs all the spell vars, renames some of them
- removes some useless vars
- start with pointed spells, as they're easy

MrMelbert committed:
kill proc_holder pt2
- kill a buncha vars and replace it with flags
- convert a ton over
- general code improvements

MrMelbert committed:
kill proc_holders pt3
- convert a good few more spells
- rename some signals
- handle statpanel
- better docs

MrMelbert committed:
kiill proc_holder pt4:
- restructure the file system of action.dm, separating a good amount of item actions and miscellaneous garbage into files where they belong slightly better. Also splits off item actions, cooldown actions, innate actions, etc. into their own files, overlal making it much better to work with
- converts touch attacks to actions
- converts blood crawl, jaunt subtype

MrMelbert committed:
kills proc_holder pt5
- clears up some icon issues so all the currently converted pages don't have errors
- shapeshift
- some more action cleanup

MrMelbert committed:
kills proc_holder pt5.5:
- some documentation
- reworks feedback to prevent oversight with teleports and stuff

MrMelbert committed:
kills proc_holder pt6:
- converted cult spells
- converted magic missile
- converted mime spells
- chipped away at the errors
- removed some vars which were too general, replaced them with more locally applicable vars. for example "range" which could mean "projectile range" or "aoe radius" or whatever - instead of having a broad net which everyone applies to in a confusing matter, instead lets each spell delegate on their own.
- merged magic/spell and magic/aoe, as the comment intended
- more unified behavior for spell levelling

MrMelbert committed:
kill proc_holders pt 6.5:
- replacing a buncha old proc_holders that have been updated to reduce some errors. sub 900 baby

MrMelbert committed:
kills proc_holder pt 6.75:
- minor fixes

MrMelbert committed:
kills proc_holder pt7:
- cuts down on some errors
- refactors some wiz events

MrMelbert committed:
kills proc_holder pt 7.5:
- malf ranged modules
- some minor errors

MrMelbert committed:
kills proc_holder pt 7.75:
- mor eminor error handling, cleaning up changes

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
kill proc_holder pt8:
- refactors spell book
- refactors spell implant
- some more minor error fixing

MrMelbert committed:
kill proc_holder pt 8.5:
- scan ability

MrMelbert committed:
Adds some robust documentation

MrMelbert committed:
kill proc_holder pt9:
- converts some / most mutations over

MrMelbert committed:
kill proc_holder pt10:
- sort out all the granters
- refactor them slightly
- fix some compile errors

MrMelbert committed:
Some set-unset sanity - going to need to test removing Share()

MrMelbert committed:
Removes transfer actions. It doesn't seem to do anything.
- Transfer_actions was called when current = new_character so locially speaking the early return in Grant() should cause it to NOOP. Test this in the future though

MrMelbert committed:
Removes sharing from actions, docs actions better

MrMelbert committed:
Some better documentation for spell and spell components

MrMelbert committed:
Kills proc_holder pt11:
- Finally finishes ALL THE SPELLS IN THE SPELL FOLDER
- Fixes some more errors

MrMelbert committed:
kills proc_holder pt11.5:
- minor error fixing and sanity

MrMelbert committed:
Method of sharing actions. Can be improved  in the future, needs testing

MrMelbert committed:
Implements a way to update the stat panel entry for a spell. Also gets rid of VV stuff, as you can update the bigflags directly in VV now.

MrMelbert committed:
Curse of madness bug I put in.

MrMelbert committed:
kills proc_holder pt12:
- sub 500 errors!
- converts cytology mobs
- converts and refactors spiders slightly
- some minor fixing around the place as usual

MrMelbert committed:
kill proc_holder pt13
- Finishes heretic spells
- Sub 300 errors!
- some touch refactoring to account for mansus grasp

MrMelbert committed:
kills proc_holder pt14:
- revenant
- minor bugfixing for heretic stuff

MrMelbert committed:
kills proc_holder pt14.5:
- some missed stuff for revenant + heretic

MrMelbert committed:
kills proc_holder pt15:
- alien abilities
- more minor fixing
- sub 100 errors. The end is nigh

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
kill proc_holder pt16? 17:
- Finishes cult spells
- sub 50 errors!
- refactors the way charge works
- renames / moves some signals

MrMelbert committed:
kills proc_holder pt final:
- sdql spells
- no more errors!

MrMelbert committed:
Bugfixes round 1

MrMelbert committed:
Various bugfixing
- documentation done
- give spell works
- can cast spell gives feedback conditionally
- is available takes into account casting ability

MrMelbert committed:
Some accidental reversions + fixes

MrMelbert committed:
Unit tests

MrMelbert committed:
Completely refactors jaunting
- All bloodcrawling is now handled on the action itself instead of across various living procs
- slaughter demons have their own blood crawls
- jaunting dummies don't have side effects on destroy() anymore

MrMelbert committed:
Wizard spell logging and even more refactoring
- minor fixing
- wizard scrolls use teleport actions

MrMelbert committed:
Void pull fix

MrMelbert committed:
More bugfixes, primarily for pointed projectiles

MrMelbert committed:
Changed some action signals

MrMelbert committed:
Some spell corrections, mime spells update(?)

MrMelbert committed:
Some unit test fixing

MrMelbert committed:
Revert -> reset

MrMelbert committed:
Whoops. Redo that one

MrMelbert committed:
Minor documentation and code tweaks

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Post merge fixes

MrMelbert committed:
debug text

MrMelbert committed:
Sanity checks around the place, namely for owner

MrMelbert committed:
some bug fixing + lemon will yell at me for this

MrMelbert committed:
comment for later.

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Post merge fixes

MrMelbert committed:
Comment

MrMelbert committed:
Mansus Grasp silliness

MrMelbert committed:
Optimizes aoe spells

MrMelbert committed:
Retires the mime element for now

MrMelbert committed:
Cleans up a few spells

MrMelbert committed:
Mime spell fixes, granter fixes

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
These shouldn't have returned

MrMelbert committed:
Split up the tests

MrMelbert committed:
Touch tweaks

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Converting over the heretic changes

MrMelbert committed:
Fixes for spells

MrMelbert committed:
c

MrMelbert committed:
Tweaks before more refactoring

MrMelbert committed:
Updates telepathy

MrMelbert committed:
Personality commune, too

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Post merge cleanups

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Comment.

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
whisper, thanks novaa

MrMelbert committed:
Bugfixing

MrMelbert committed:
Jaunt tweaks + abstract stuff

MrMelbert committed:
More minor fixes

MrMelbert committed:
Update mouse pointer

MrMelbert committed:
Jaunting minor refactoring, more bugfixing

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Updatepaths script

MrMelbert committed:
Map reset

MrMelbert committed:
Run updatepaths

MrMelbert committed:
Pre pr changes

bruhlookatthisdood commented:
very based, goodies.

MrMelbert committed:
Fixes some linting errors

MrMelbert committed:
Yet again fire breath linters

MrMelbert committed:
Pleases some proc validity lints

LemonInTheDark commented:
Brother this is the shit

MrMelbert committed:
More minor bugfixing

MrMelbert committed:
Slight adjustment to heretic stuff, sanity checks

ATH1909 commented:
> qol: Genetics mindread is now a pointed ability.
> qol: The mimery book now grants you a vow of silence action, like the guide to advanced mimery.

❤️ 

> balance: Only mindswap can be cast if the wizard is a brain. Most spells didn't work, anyways.
> balance: PAIs and carded AIs can no longer cast spells. Most spells didn't work, anyways.

sad!

could you at least let mind transfer work if you're a pAI/carded AI? it'd be pretty lame to mind transfer into a pAI and then get stuck in their body for an arbitrary reason, and being carded as a wizard AI sounds like a situation you might want to mind transfer out of.

MrDoomBringer commented:
changelog needs more detail

MrMelbert committed:
Spider cleanup + Some more busywork

MrMelbert commented:
> is there any particular reason for this?

so, currently, spell books keep a hard ref to the mob that owns it 
if the mob gets qdeleted, the mob will hard delete, and also it'll clear the reference which will allow anyone to re-link after the gc cleans up the reference

so when I cleared there reference to prevent hard-delling it allowed anyone to instantly link to the book, which seems wrong, so I just qdel'd the book with it + some flavor.

VladinXXV commented:
I think regal pipegun and trash cannon are already pretty rarely made/used, and that's considering most people share the recipe book after finding it. I think they should be unlimited use, just like the other recipe granters.

MrMelbert commented:
> I think regal pipegun and trash cannon are already pretty rarely made/used, and that's considering most people share the recipe book after finding it. I think they should be unlimited use, just like the other recipe granters.

![image](https://user-images.githubusercontent.com/51863163/168956491-7ecece8e-a86b-4146-8799-0ff18de8c6b1.png)

*shrug
I'm fine with changing them, I'll have to ask the original coders

Palpetinus commented:
> You can now drop the hand to immediately cancel touch spells.

On wizards, nice. Does this also affect Heretics or rather, was this also changed for Heretics? It's been baffling to me how on cult you can just "drop" any spell but on Heretics I have to tediously click the button again

timothymtorres commented:
This PR is a reckoning for magic code that needed to be badly refactored.

B4CKU commented:
> qol: Genetics mindread is now a pointed ability.

holy based

MrMelbert committed:
Some more general cleanup

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert commented:
> > You can now drop the hand to immediately cancel touch spells.
> 
> On wizards, nice. Does this also affect Heretics' Grasp of Mansus or rather, was this also changed for Heretics? It's been baffling to me how on cult you can just "drop" any spell but on Heretics I have to tediously click the button again

Yes, the mansus grasp spell is a touch spell, so it also benefits from being able to drop it to cancel the cast.

MrMelbert committed:
Compile

MrMelbert committed:
Works on fixing the stat panel

MrMelbert committed:
Statpanel works again

MrMelbert committed:
Compiles

MrMelbert committed:
I didn't save

MrMelbert committed:
This actually isn't necessary at all

MrMelbert committed:
Whoops

MrMelbert committed:
Renames and moves some unit tests

MrMelbert committed:
Granter bugfixing + feedback tweaks

magatsuchi commented:
![engineer-o](https://user-images.githubusercontent.com/88991542/169386569-6b01cd4c-6f68-401c-b357-9ed5bb73343a.gif)

THAT DIFF THO

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Readds cult hallucination with some meh-code

MrMelbert committed:
Some missed things. Hastily got it up

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
maps hard reset

MrMelbert committed:
re-apply updatepaths

MrMelbert committed:
re, RE hard reset maps

MrMelbert committed:
re, RE apply updatepaths

MrMelbert committed:
Beforecast refactor, to fix some issues

MrMelbert committed:
Fixes benos

MrMelbert committed:
Maybe fixes a heretic bug. Test this later

MrMelbert committed:
See above

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Testing fixes

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

SuperNovaa41 commented on code/datums/actions/action.dm:
```diff
@@ -0,0 +1,254 @@
+		qdel(src)
+
+/// Grants the action to the passed mob, making it the owner
+/datum/action/proc/Grant(mob/M)
```
while you're here

SuperNovaa41 commented on code/datums/actions/action.dm:
```diff
@@ -0,0 +1,254 @@
+		current_button.add_overlay(mutable_appearance(icon_icon, button_icon_state))
+		current_button.button_icon_state = button_icon_state
+
+//Give our action button to the player
```
proper doc

SuperNovaa41 commented on code/datums/actions/action.dm:
```diff
@@ -0,0 +1,254 @@
+	LAZYOR(viewer.actions, src) // Move this in
+	ShowTo(viewer)
+
+//Adds our action button to the screen of a player
```
proper doc

SuperNovaa41 replied:
same for the rest of this file i suppose

SuperNovaa41 commented on code/datums/actions/innate_action.dm:
```diff
@@ -0,0 +1,58 @@
+//Preset for general and toggled actions
```
doc comment

SuperNovaa41 commented on code/datums/actions/items/beserk.dm:
```diff
@@ -0,0 +1,19 @@
+
+/datum/action/item_action/berserk_mode/Trigger(trigger_flags)
+	if(istype(target, /obj/item/clothing/head/hooded/berserker))
+		var/obj/item/clothing/head/hooded/berserker/berzerk = target
```
spelled berserk wrong...

SuperNovaa41 commented on code/datums/actions/items/beserk.dm:
```diff
@@ -0,0 +1,19 @@
+	background_icon_state = "bg_demon"
+
+/datum/action/item_action/berserk_mode/Trigger(trigger_flags)
+	if(istype(target, /obj/item/clothing/head/hooded/berserker))
```
since you're returning at the end of this proc regardless, couldn't you just do uhh
```
if (!istype())
    return ..()
....
```

SuperNovaa41 commented on code/datums/actions/items/cult_dagger.dm:
```diff
@@ -0,0 +1,38 @@
+	background_icon_state = "bg_demon"
+	default_button_position = "6:157,4:-2"
+
+/datum/action/item_action/cult_dagger/Grant(mob/M)
```
var

SuperNovaa41 commented on code/datums/actions/items/organ_action.dm:
```diff
@@ -0,0 +1,25 @@
+	check_flags = AB_CHECK_CONSCIOUS
+
+/datum/action/item_action/organ_action/IsAvailable()
+	var/obj/item/organ/I = target
```
.

SuperNovaa41 commented on code/datums/actions/items/set_internals.dm:
```diff
@@ -0,0 +1,12 @@
+		return
+	if(!iscarbon(owner))
+		return
+	var/mob/living/carbon/C = owner
```
.

SuperNovaa41 commented on code/datums/actions/items/stealth_box.dm:
```diff
@@ -0,0 +1,55 @@
+	owner.forceMove(box)
+	owner.playsound_local(box, 'sound/misc/box_deploy.ogg', 50, TRUE)
+
+/datum/action/item_action/agent_box/Grant(mob/M)
```
.

SuperNovaa41 commented on code/datums/actions/items/toggles.dm:
```diff
@@ -0,0 +1,143 @@
+	name = "Toggle Welding Screen"
+
+/datum/action/item_action/toggle_welding_screen/Trigger(trigger_flags)
+	var/obj/item/clothing/head/hardhat/weldhat/H = target
```
.

SuperNovaa41 commented on code/datums/actions/items/vortex_recall.dm:
```diff
@@ -0,0 +1,16 @@
+		to_chat(owner, span_notice("[target] fizzles uselessly."))
+		return
+	if(istype(target, /obj/item/hierophant_club))
+		var/obj/item/hierophant_club/H = target
```
.

SuperNovaa41 commented on code/datums/actions/mobs/language_menu.dm:
```diff
@@ -0,0 +1,13 @@
+	if(!..())
+		return FALSE
+	if(ismob(owner))
+		var/mob/M = owner
+		var/datum/language_holder/H = M.get_language_holder()
```
.

SuperNovaa41 commented on code/datums/mutations/antenna.dm:
```diff
@@ -46,62 +46,77 @@
+	//this whole loop puts the read-ee's say logs into say_log in an easy to access way
+	for(var/log_type in log_source)
+		var/nlog_type = text2num(log_type)
+		if(nlog_type & LOG_SAY)
```
early continue?

SuperNovaa41 commented on code/datums/mutations/autotomy.dm:
```diff
@@ -0,0 +1,42 @@
+	name = "Autotomy"
+	desc = "Allows a creature to voluntary discard a random appendage."
+	quality = POSITIVE
+	text_gain_indication = "<span class='notice'>Your joints feel loose.</span>"
```
cant you use a span macro for this

SuperNovaa41 commented on code/datums/mutations/fire_breath.dm:
```diff
@@ -0,0 +1,92 @@
+	// Try to set us to our original direction after, so we don't end up backwards.
+	living_cast_on.setDir(original_dir)
+
+// Makes the cone shoot out into a 3 wide column of flames.
```
```suggestion
/// Makes the cone shoot out into a 3 wide column of flames.
```

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
review

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Maps reset

MrMelbert committed:
Alien fixes?

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Honorbound runtime

MrMelbert committed:
Fixes chameleon items
- Not an ideal fix, and a better fix would be to cut down on all this copy+paste stuff the chameleon kit items does, but that's a much bigger PR

MrMelbert committed:
A buncha item action fixes.

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Item actions grr

SuperNovaa41 commented on code/datums/mutations/fire_breath.dm:
```diff
@@ -0,0 +1,94 @@
+	to_modify.cone_levels += 2  // Cone fwooshes further, and...
+	to_modify.self_throw_range += 1 // the breath throws the user back more
+
+	return .
```
i get you're doing this for readabiliy sake, but its consistent with the other returns in this proc

SuperNovaa41 commented on code/datums/mutations/fire_breath.dm:
```diff
@@ -0,0 +1,94 @@
+	if(. & SPELL_CANCEL_CAST)
+		return
+
+	if(iscarbon(cast_on))
```
early return

SuperNovaa41 commented on code/datums/mutations/holy_mutation/honorbound.dm:
```diff
@@ -201,72 +201,118 @@
+	/// A ref to our owner's honorbound mutation
+	var/datum/mutation/human/honorbound/honor_mutation
+	/// The declaration that's shouted in invocation. Set in New()
+	var/declaration = "By the divine light of my diety, you are an evil of this world that must be wrought low!"
```
```suggestion
	var/declaration = "By the divine light of my deity, you are an evil of this world that must be wrought low!"
```

SuperNovaa41 commented on code/datums/status_effects/neutral.dm:
```diff
@@ -103,13 +103,12 @@
-			spell.charge_counter = spell.charge_max
-			spell.recharging = FALSE
-			spell.update_appearance()
+		for(var/datum/action/cooldown/spell/spell in rewarded.actions)
```
as anything here?

SuperNovaa41 commented on code/game/objects/effects/forcefields.dm:
```diff
@@ -1,12 +1,13 @@
 	density = TRUE
 	can_atmos_pass = ATMOS_PASS_DENSITY
-	var/timeleft = 300 //Set to 0 for permanent forcefields (ugh)
+	/// Set to 0 for permanent forcefields (ugh)
```
this is a bad doc comment its better left as `//`

SuperNovaa41 commented on code/modules/admin/verbs/SDQL2/SDQL_spells/spell_edit_menu.dm:
```diff
@@ -21,79 +21,72 @@ GLOBAL_LIST_INIT_TYPED(sdql_spells, /obj/effect/proc_holder/spell, list())
+		"max_targets",
+		"aoe_radius",
+
+		//"requires_human",
+		//"requires_non_abstract",
+		//"requires_unphased",
+		//"requires_no_antimagic",
+		//"can_cast_on_centcom",
+		//"requires_wizard_garb",
+		//"requires_conscious",
```
?

SuperNovaa41 reviewed (changes requested):
Taking a break because githubs syntax highlighting broke and reading this as plain text is agony

MrMelbert committed:
Clean up modify()

MrMelbert committed:
Review

MrMelbert committed:
Forcefields

MrMelbert committed:
SDQL spells are dead

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Fixes

MrMelbert committed:
Fixes vatbeast

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Removes final todo?

MrMelbert committed:
Thermal vision fix

MrMelbert committed:
Steel

MrMelbert committed:
Fine, I give up

MrMelbert committed:
Teleport fixes and other oddities

MrMelbert committed:
Thermal fix again

MrMelbert committed:
Shit

MrMelbert committed:
This is cleaner

MrMelbert committed:
Cleaner?

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Uniformity

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Summon item runtime error

MrMelbert committed:
Fixes modsuit modules

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

jlsnow301 commented on code/datums/actions/item_action.dm:
```diff
@@ -0,0 +1,33 @@
+		// If set, use the custom icon that we set instead
+		// of the item appearence
+		..()
+	else if((target && current_button.appearance_cache != item_target.appearance) || force) //replace with /ref comparison if this is not valid.
```
this is probably fine, but is it ok if you use else \newline if since "else if" isnt real here? 
or even something like
```
        else 
	    if(!target || current_button.appearance_cache == item_target.appearance) || !force)
	        return
	    ....
```

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Fixes agent IDs

MrMelbert committed:
Small op for necropotence bolt

MrMelbert committed:
Projectile stuff

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Missed a parenthesis

MrMelbert committed:
Switches a hardref to a ref

MrMelbert committed:
Better logging

MrMelbert committed:
Oops, this was leftover.

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
UI stuff

MrMelbert committed:
Lints?

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Fix the mime box

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Maps reset

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Jaunt effect tweaks

MrMelbert committed:
Correct linters

MrMelbert committed:
Missed one, because it did not save.

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

SuperNovaa41 reviewed (commented):
Going to look at the actual base spell stuff again, but all of the small change files + all the redone spells look good

MrMelbert committed:
Merge branch 'master' into kill_proc_holder

MrMelbert committed:
Fixes implant actions

MrMelbert committed:
Review

SuperNovaa41 commented on code/modules/instruments/items.dm:
```diff
@@ -215,6 +215,17 @@
+
+/datum/action/item_action/instrument/Trigger(trigger_flags)
+	if(istype(target, /obj/item/instrument))
+		var/obj/item/instrument/I = target
```
all good i couldn't tell if you added this or if it was copy and pasted, hard to tell with just diffs

LemonInTheDark commented on code/modules/spells/spell.dm:
```diff
@@ -1,609 +1,454 @@
-		if(nonabstract_req && (isbrain(user) || ispAI(user)))
-			to_chat(user, span_warning("This spell can only be cast by physical beings!"))
+
+		if((!(spell_requirements & SPELL_CASTABLE_AS_BRAIN)) && isbrain(owner))
```
```suggestion
		if(!(spell_requirements & SPELL_CASTABLE_AS_BRAIN) && isbrain(owner))
```

LemonInTheDark commented on code/modules/spells/spell.dm:
```diff
@@ -1,609 +1,454 @@
+	// If the spell requires the user has no antimagic equipped, and they're holding antimagic
+	// that corresponds with the spell's antimagic, then they can't actually cast the spell
+	if((spell_requirements & SPELL_REQUIRES_NO_ANTIMAGIC) && !owner.can_cast_magic(antimagic_flags))
+		if(feedback)
+			to_chat(owner, span_warning("Some form of antimagic is preventing you from casting [src]!"))
 		return FALSE
 
-	if(ishuman(user))
+	if((!(spell_requirements & SPELL_CASTABLE_WHILE_PHASED)) && HAS_TRAIT(owner, TRAIT_MAGICALLY_PHASED))
```
you don't need this second () around the checks

LemonInTheDark commented on code/modules/spells/spell.dm:
```diff
@@ -1,609 +1,454 @@
-				user.whisper(replacetext(invocation," ","`"))
-		if(INVOCATION_EMOTE)
-			user.visible_message(invocation, invocation_emote_self) //same style as in mob/living/emote.dm
+// The actual cast chain occurs here, in Activate().
+// You should generally not be overriding or extending Activate() for spells.
+// Defer to any of the cast chain procs instead.
+/datum/action/cooldown/spell/Activate(atom/cast_on)
```
this isn't your fault but the way we do this makes me sad

LemonInTheDark commented on code/modules/spells/spell.dm:
```diff
@@ -1,609 +1,454 @@
-	if(cast_check())
-		choose_targets()
-	return 1
+	var/sig_return = SEND_SIGNAL(src, COMSIG_SPELL_BEFORE_CAST, cast_on)
+	if(owner)
+		sig_return |= SEND_SIGNAL(owner, COMSIG_MOB_BEFORE_SPELL_CAST, src, cast_on)
```
why do we need both? ease of use?

LemonInTheDark replied:
why both? ease of use?

LemonInTheDark commented on code/modules/spells/spell.dm:
```diff
@@ -1,609 +1,454 @@
+	SHOULD_CALL_PARENT(TRUE)
+
+	SEND_SIGNAL(src, COMSIG_SPELL_AFTER_CAST, cast_on)
+	if(owner)
```
early return?

LemonInTheDark commented on code/modules/spells/spell.dm:
```diff
@@ -1,609 +1,454 @@
+	var/mob/living/living_owner = owner
+	if(invocation_type == INVOCATION_EMOTE && HAS_TRAIT(living_owner, TRAIT_EMOTEMUTE))
+		if(feedback)
+			to_chat(owner, span_warning("You can't get form to invoke [src]!"))
```
get form?

LemonInTheDark commented on code/modules/spells/spell.dm:
```diff
@@ -1,609 +1,454 @@
+	name = "[spell_title][initial(name)]"
+	UpdateButtons()
+
+/// Debug proc for seeing at a glance what all spells have as set requirements
+/// Probably remove before merge, idk it might be useful in the future as a debug verb.
+/proc/debug_spell_requirements()
+	for(var/datum/action/cooldown/spell/spell as anything in typesof(/datum/action/cooldown/spell))
```
if you keep it, make it a debug verb (if not already)
also sum the text rather then just spamming messages

MrMelbert committed:
Final review?

MrMelbert committed:
Final touches

MrMelbert committed:
Name unit test

MrMelbert committed:
Bonus unit test

SuperNovaa41 commented on code/game/objects/items/implants/implant.dm:
```diff
@@ -16,14 +18,17 @@
 	INVOKE_ASYNC(src, .proc/activate, "action_button")
 
+/obj/item/implant/item_action_slot_check(slot, mob/user)
+	return (user == imp_in)
```
```suggestion
	return user == imp_in
```
if lemon gets to nitpick so do it

SuperNovaa41 commented on code/game/objects/items/implants/implant_spell.dm:
```diff
@@ -1,36 +1,56 @@
 
+/obj/item/implant/spell/Initialize(mapload)
+	. = ..()
+	if(spell_type)
```
early return?

SuperNovaa41 reviewed (changes requested):
Ok i think this is ready

MrMelbert committed:
Final nitpicks

SuperNovaa41 commented:
Rip off the bandaid amirite
